### PR TITLE
Allow prepending access log handler for WASM filter

### PR DESCRIPTION
**Pull request by sunaydagli**
(Duplicate of #35910, mistakenly did not follow DCO and used wrong fixing command)

Commit Message: Allow prepending access log handler for WASM filter
Additional Description:
Risk Level: Low
Testing: Integration
Docs Changes:
Release Notes:
Platform Specific Features:
Runtime guard: envoy.reloadable_features.prepend_access_log_handler
Fixes https://github.com/envoyproxy/envoy/issues/30859

This PR is in response to the approaches considered in #https://github.com/envoyproxy/envoy/pull/35595 and https://github.com/envoyproxy/envoy/pull/35742 in order to solve https://github.com/envoyproxy/envoy/issues/30859. The recommended approach in these discussions was to override onStreamComplete callback, which may make sense for most filters. However, the WASM filter is a unique case in which doing this method would be more of a work-around hack as the [filter itself](https://github.com/envoyproxy/envoy/tree/ac402243922317ae7ca8c601e6cdb243a4f726ad/source/extensions/filters/http/wasm) does not inherit StreamFilter, and while the common WASM section does, it cannot utilize the onStreamComplete from its individual filters and thus is missing onStreamComplete, instead using [onDone](https://github.com/envoyproxy/envoy/blob/ac402243922317ae7ca8c601e6cdb243a4f726ad/source/extensions/common/wasm/context.cc#L1536) as a proxy for it. As a result, implementing onStreamComplete here would lead to onLog being called within the onStreamComplete method instead of the log() method, which doesn't seem to make the most sense for readability.

As such, we propose to continue the initial decided implementation of prepending the access loggers in order to maintain the execution order. This solution does seem to address this specific problem as described in the issue, so that AccessLogger doesn't see the dynamicMetadata until WASM finishes writing to it. This would be a case specific to WASM because of the details described above. Another WASM implementation is dependent on this issue being fixed, would welcome feedback or approval on this PR.

### CODE REVIEW

**Comment by sunaydagli** [comment]
Apologies, I had to open a new PR since I mistakenly used the wrong command to fix GCO issues, please refer to #35910 for previous discussions.

@spacewander, I did consider using the approach you suggested, but since there may be more immediate feature changes to be made, this solution does answer the prior issue in #30859. That being said, if the wasm access log is the more appropriate solution after this is rolled out, we can definitely change it to that approach. I figure since this feature is only used for WASM, it should be safe enough, especially now that it is runtime guard.

@wbpcode thank you for the feedback, I added a runtime guard and added information to the changelog.

**Comment by sunaydagli** [comment]
FYI: tests failing are the same flakey tests to be addressed in this [PR](https://github.com/envoyproxy/envoy/issues/35523).

**Comment by wbpcode** [comment]
My previous comment is means that we can execute all loggers that be added by the filters before the access loggers. By that way, the order of loggers (added by filters) will still keep consistence with `http_filters`.

I need to add a new `prependAccessLogHandler` API.

And I didn't get this point:

>  As a result, implementing onStreamComplete here would lead to onLog being called within the onStreamComplete method instead of the log() method, which doesn't seem to make the most sense for readability.

Why? The `onStreamComplete` should be only used to tell the wasm filter the current stream is end. And wasm filter could do something in the method.

**Comment by sunaydagli** [comment]
> My previous comment is means that we can execute all loggers that be added by the filters before the access loggers. By that way, the order of loggers (added by filters) will still keep consistence with `http_filters`.
> 
> we needn't to add a new `prependAccessLogHandler` API.
> 
> And I didn't get this point:
> 
> > As a result, implementing onStreamComplete here would lead to onLog being called within the onStreamComplete method instead of the log() method, which doesn't seem to make the most sense for readability.
> 
> Why? The `onStreamComplete` should be only used to tell the wasm filter the current stream is end. And wasm filter could do something in the method.

Could you clarify what you mean by executing all loggers that are added by the filters before the access loggers? As in, how would this be done?

For the second point, the WASM context.cc does not override onStreamComplete, because the underlying filter in this [directory](https://github.com/envoyproxy/envoy/tree/7ffff7aa7390581bb02cbc6ebbbfa267fc7e1945/source/extensions/filters/http/wasm) only generates the filter from a config, and does not actually define the filter with inheriting StreamFilter or StreamFilterBase (from which onStreamComplete is overrode from). The context.cc has a method, [onDone()](https://github.com/envoyproxy/envoy/blob/7ffff7aa7390581bb02cbc6ebbbfa267fc7e1945/source/extensions/common/wasm/context.cc#L194), which seems to be a proxy for this in other methods, as even if the onStreamComplete is implemented, it would not be able to utilize the WASM filter's onStreamComplete in the context.cc implementation. Therefore that's why I thought prepending specifically for WASM would address this issue.

**Comment by wbpcode** [comment]
> as even if the onStreamComplete is implemented, it would not be able to utilize the WASM filter's onStreamComplete in the context.cc implementation.

Sorry, I still didn't get it. In my mind, the WASM extension is constructed by two parts: 1. wrapper in the Envoy that is implemented in C++ (Context and code in the proxy-wasm-cpp-host). 2. actual extension logic in the WASM bytes code which could be implemented in any language.

You only need to extend both to implement/provide the `onStreamComple` or similar things, then it should be work.

Feel free to point it out if I am wrong.

**Comment by wbpcode** [comment]
> Could you clarify what you mean by executing all loggers that are added by the filters before the access loggers? As in, how would this be done?

Now, the access loggers (from configuration) and filter loggers (from filters) are store in a single list. And the access loggers will be added first.

So, the finally execution order is the access loggers first.

But it's possible to change it by keeping them in two containers and execute the filter loggers first.

**Comment by sunaydagli** [comment]
> > as even if the onStreamComplete is implemented, it would not be able to utilize the WASM filter's onStreamComplete in the context.cc implementation.
> 
> Sorry, I still didn't get it. In my mind, the WASM extension is constructed by two parts: 1. wrapper in the Envoy that is implemented in C++ (Context and code in the proxy-wasm-cpp-host). 2. actual extension logic in the WASM bytes code which could be implemented in any language.
> 
> You only need to extend both to implement/provide the `onStreamComple` or similar things, then it should be work.
> 
> Feel free to point it out if I am wrong.

I believe what you are saying is correct for wasm, but this is for the wasm filter as well, which would be located in this [directory](https://github.com/envoyproxy/envoy/tree/45dc2810e1216576eaebe047f4efe6b5a63eb0c9/source/extensions/filters/http/wasm). What you are saying to implement/provide onStreamComplete in this area is not possible without significant wasm/http structure changes as if you go through envoy/source/extensions/filters/http, only 1 or 2 have onStreamComplete, and these are usually helper filters that contain or reference another (an example is /[match_delegate](https://github.com/envoyproxy/envoy/tree/45dc2810e1216576eaebe047f4efe6b5a63eb0c9/source/extensions/filters/http/match_delegate) in that directory) and that allows using the other nested filter's onStreamComplete. If this directory is not to be used for solving the issue at hand, then please let me know. That is why I think this prepend solution works in particular for wasm.


> > Could you clarify what you mean by executing all loggers that are added by the filters before the access loggers? As in, how would this be done?
> 
> Now, the access loggers (from configuration) and filter loggers (from filters) are store in a single list. And the access loggers will be added first.
> 
> So, the finally execution order is the access loggers first.
> 
> But it's possible to change it by keeping them in two containers and execute the filter loggers first.

To clarify if I am understanding this suggestion properly, do you mean changing the order of this [function](https://github.com/envoyproxy/envoy/blob/45dc2810e1216576eaebe047f4efe6b5a63eb0c9/source/extensions/common/wasm/context.cc#L1091) to have access loggers first? I am unsure where the two containers portion comes into play, would it just mean having two callback lists?

Thanks again for your time for this I appreciate it!

**Comment by wbpcode** [comment]
I mean you can split the access_log_handlers_ two to different contains. One for filters loggers, one for access loggers. And always iterate the filters loggers first.

https://github.com/envoyproxy/envoy/blob/cbfea5b91657c09021df178c5d7a8495fd7f70fe/source/common/http/filter_manager.h#L1073

File: changelogs/current.yaml [hunk 1]
```diff
@@ -39,6 +39,14 @@ behavior_changes:
   change: |
     Added new tag extraction so that scoped rds stats have their :ref:'scope_route_config_name
     <envoy_v3_api_msg_config/route/v3/scoped_route>' and stat prefix extracted.
+- area: wasm
+  change: |
+    Added prependAccessLogHandler in HTTP filter to be used by WASM in order to fix
```
Comment (phlax) [inline changelogs/current.yaml:44]:
```suggestion
    Added ``prependAccessLogHandler`` in HTTP filter to be used by WASM in order to fix
```

File: changelogs/current.yaml [hunk 2]
```diff
@@ -39,6 +39,14 @@ behavior_changes:
...
+    `issue 30859 <https://github.com/envoyproxy/envoy/issues/30859>`_, which allows
+    reversing adding the order of access loggers in order to provide correct dynamic metadata flow
```
Comment (phlax) [inline changelogs/current.yaml:46]:
```suggestion
    reversing the order access loggers are added in order to provide correct dynamic metadata flow
```

File: changelogs/current.yaml [hunk 3]
```diff
@@ -39,6 +39,14 @@ behavior_changes:
...
+    without inheriting onStreamComplete(). This solution is exclusively used by WASM, and may be
```
Comment (phlax) [inline changelogs/current.yaml:47]:
```suggestion
    without inheriting ``onStreamComplete()``. This solution is exclusively used by WASM, and may be
```

File: changelogs/current.yaml [hunk 4]
```diff
@@ -39,6 +39,14 @@ behavior_changes:
...
+    changed to an updated WASM access logger if additional features deems it necessary. This behavior can
+    be reverted by setting the runtime guard ``envoy.reloadable_features.prepend_access_log_handler`` to false.
```
Comment (phlax) [inline changelogs/current.yaml:49]:
```suggestion
    be reverted by setting the runtime guard ``envoy.reloadable_features.prepend_access_log_handler`` to ``false``.
```

**Pull request closed without merging**